packages/@aws-cdk/aws-cognito/lib/user-pool-client.ts

```diff
@@ -182,6 +183,20 @@ export interface UserPoolClientOptions {
    * @default true for new stacks
    */
   readonly preventUserExistenceErrors?: boolean;
+
+  /**
+   * Whether users registered in the user pool should be able to sign in using this client.
+   * If this is set to `true` and `identityProviders` are set, the client will only allow sign in via
+   * third-party identity providers.
+   * @default true
+   */
+  readonly allowUserPoolIdentities?: boolean;
```

Comment:
Would it be more clear if this was called `allowCognitoIdentities`? 

Why would I ever not want to set this to `true`? In that case, why I do I even have an associated `UserPool`?

Reply:
I'm also not even sure that `allowUserPoolIdentities` covers the functionality.

As far as I can tell from the way the object model seems to hold together, the identities added by a "user pool identity provider" become *part* of the pool, right? Because you register them with a pool, not with a client. As such, `allowUserPoolIdentities: false` to me would mean there is just no sign-in possible, which of course would be silly.

Here's my mental model (which might be wrong, so please tell me if it is):

```
┌────────────────────────────────────────────────┐
│                    UserPool                    │
│                                                │
│                                                │
│ ┌──────────┐    ┌──────────┐     ┌──────────┐  │
│ │          │    │          │     │          │  │
│ │ Cognito  │    │  Amazon  │     │ Facebook │  │
│ │          │    │          │     │          │  │
│ └──────────┘    └──────────┘     └──────────┘  │
└────────────────────────────────────────────────┘
                         ▲                        
                         │                        
                         │                        
          ┌─────────────────────────────┐         
          │                             │         
          │       UserPoolClient        │         
          │                             │         
          └─────────────────────────────┘         
```

As such, it feels like the default behavior of `UserPoolClient` should be to allow all identities in the pool (local or otherwise) to be able to log in, and only incidentally to override/restrict the identities that can make use of a client.

To make that happen we'd need `IUserPool` to have a list of `registeredIdentityProviders`, and you'd need to provide that list when doing a `UserPool.fromAttributes()`... might seem bad but OTOH you currently also don't have a good story around an imported userpool that already has identity providers... using `UserPoolIdentity.facebook()` would create a new one which is not what was intended...

Reply:
> Would it be more clear if this was called allowCognitoIdentities?

I like it.

> the identities added by a "user pool identity provider" become part of the pool, right? Because you register them with a pool, not with a client. As such, `allowUserPoolIdentities: false` to me would mean there is just no sign-in possible, which of course would be silly.

If a user signs up using a 3rd party social provider (i.e., clicks 'Sign In With Facebook' for the first time), the user does become part of the user pool. However, the user going forward can only use facebook to sign in.
If a user signs up directly with a user pool, a username and password is made available for his future sign in and he can only use that to sign in.

So if a user pool has `allowUserPoolIdentities` set to `false`, it means users can only sign in via enabled 3rd party identity providers and cannot have their sign in experience directly in Cognito.

Reply:
> To make that happen we'd need IUserPool to have a list of registeredIdentityProviders, and you'd need to provide that list when doing a UserPool.fromAttributes()... might seem bad but OTOH you currently also don't have a good story around an imported userpool that already has identity providers... using UserPoolIdentity.facebook() would create a new one which is not what was intended...

You suggested change feels a bit awkward.

How about something like this instead -

```ts
new UserPoolClient(this, 'Client', {
  ...
  identityProviders [ IdentityProvider.Facebook, IdentityProvider.Google, IdentityProvider.Cognito ],
});
```

with the default being `IdentityProvider.Cognito`.

Reply:
The only caveat here is that if the user specifies an `IdentityProvider` here that isn't registered with the `UserPool`, deploy will fail.

Reply:
> You suggested change feels a bit awkward.

Can you explain what about it is awkward in your opinion? Also, awkward for the user or for the implementor?

There are two steps to identity providers, as far as I can tell:

* Register zero or more IDPs with a UserPool
* Create a Client and indicate the IDPs you would like to support for this Client

I feel like making the second step optional by automatically picking "all" there is exactly the kind of thing that is in CDK's ball park, and that's why people like it. It Just Does The Right Thing™️.

The reason you are probably balking at this is because the underlying APIs make it hard to make this easy from an implementation perspective... so we kinda just fall back to "whatever CloudFormation does". Kinda feels like a shame. 

This is not a force. If you really don't want to implement that, I'm also fine with that. Just thought I saw an opportunity to raise the user comfort here.

Reply:
The awkward part comes in the adjustment to `IUserPool` that would be needed. 

The use case for someone defining the UserPool, IdP and Client in the same CDK app becomes much slicker, but for someone importing a UserPool but defining the Client becomes somewhat awkward.
They'll now not only have to import the UserPool but also the IdP, in order for their Client to work correctly. In this case, it would not be possible to enable 'Facebook' unless the IdP exists in the app (defined or imported). 

Is this an acceptable burden on the user?

Your suggestion means an additional step for the user (import the different identity providers), while mine means a bit less automatic and hence a potential deploy-time error. 
Which way should this lean?

Reply:
Actually, let me take another crack at this.

Reply:
> [...] but for someone importing a UserPool but defining the Client becomes somewhat awkward.
They'll now not only have to import the UserPool but also the IdP, in order for their Client to work correctly

Yes but if we don't put it on the Import, note that they would need to do that ANYWAY in order to make it work with the API that CloudFormation/Cognito allow us.

As a thought exercise, what would you prefer users write to make a client for an imported UserPool with a Facebook provider?

Reply:
Take a look at the latest changes. 

I've implemented the automatic recognition of identity providers from the client and enabled all registered identity providers.